airbyte-notification/src/main/java/io/airbyte/notification/SlackNotificationClient.java

```diff
@@ -0,0 +1,125 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.notification;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMap.Builder;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.commons.map.MoreMaps;
+import io.airbyte.commons.resources.MoreResources;
+import io.airbyte.commons.yaml.Yamls;
+import io.airbyte.config.SlackNotificationConfiguration;
+import java.io.IOException;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.Map;
+import java.util.Map.Entry;
+import org.apache.logging.log4j.util.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Notification client that uses Slack API for Incoming Webhook to send messages.
+ *
+ * This class also reads a resource YAML file that defines the template message to send.
+ *
+ * It is stored as a YAML so that we can easily change the structure of the JSON data expected by
+ * the API that we are posting to (and we can write multi-line strings more easily).
+ *
+ * For example, slack API expects some text message in the { "text" : "Hello World" } field...
+ *
+ * It is also a template as the content of the message can be customized with basic variables that
+ * are built in the metadata map. The templating is very basic for now: it is just looking for keys
+ * from the metadata map surrounded by '<' '>' characters in the string
+ */
+public class SlackNotificationClient implements NotificationClient {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SlackNotificationClient.class);
+
+  private final HttpClient httpClient = HttpClient.newBuilder()
+      .version(HttpClient.Version.HTTP_2)
+      .build();
+  private final SlackNotificationConfiguration config;
+  private final String connectionPageUrl;
+
+  public SlackNotificationClient(final String connectionPageUrl, final SlackNotificationConfiguration config) {
+    this.connectionPageUrl = connectionPageUrl;
+    this.config = config;
+  }
+
+  @Override
+  public boolean notifyJobFailure(final Map<String, Object> jobData) throws IOException, InterruptedException {
+    if (config.getOnFailure()) {
+      return notify(renderJobData(jobData, "failure_slack_notification_template.yml"));
+    }
+    return true;
+  }
+
+  private String renderJobData(final Map<String, Object> jobData, final String templateFile) throws IOException {
+    final Map<String, Object> metadata = MoreMaps.merge(jobData, generateMetadata(jobData));
+    final JsonNode template = Yamls.deserialize(MoreResources.readResource(templateFile));
+    String data = Jsons.toPrettyString(template);
+    for (Entry<String, Object> entry : metadata.entrySet()) {
+      final String key = "<" + entry.getKey() + ">";
```

Comment:
is it worth it to do templating this way as opposed to just using `String.format()`? It feels like we're reinventing a templating protocol here.

Reply:
I went this way with the thought that we could leverage some real templating engine/library afterward to replace this basic approach. 

Having the metadata map would also maybe open the possibility for users to configure the template message in their preferences in the future.

but maybe that's too soon and we can just make something simpler for now